const ClientError = require("../../exceptions/ClientError");

class CompaniesHandler {
  constructor(service, validator) {

    this._service = service;
    this._validator = validator;

    this.postCompanyHandler = this.postCompanyHandler.bind(this);
    this.getAllCompany = this.getAllCompany.bind(this);

  }


  async getAllCompany(request, h) {
    try {
      const data = await this._service.allCompany();

      return {
        status: "success",
        data: data
      };

    } catch (error) {
      if (error instanceof ClientError) {
        const response = h.response({
          status: "fail",
          message: error.message,
        });
        response.code(error.statusCode);
        return response;
      }

      // server ERROR!
      const response = h.response({
        status: "error",
        message: "Maaf, terjadi kegagalan pada server kami.",
      });
      response.code(500);
      console.error(error);
      return response;
    }
  }

  async postCompanyHandler(request, h) {
    try {
      const { name,username,confPassword , email,address, password } = request.payload;

      // return h.response({
      //   status: request.payload,
      // });

      this._validator.validateCompanyPayload(request.payload);

      const userId = await this._service.addCompany({ name,username,confPassword , email,address, password });


      const response = h.response({
        status: "success",
        message: "User berhasil ditambahkan",
        data: {
          userId,
        },
      });
      response.code(201);
      return response;


    } catch (error) {
      if (error instanceof ClientError) {
        const response = h.response({
          status: "failed",
          message: error.message,
        });
        response.code(error.statusCode);
        return response;
      }

      // Server ERROR!
      const response = h.response({
        status: "error",
        message: "Maaf, terjadi kegagalan pada server kami.",
      });
      response.code(500);
      console.error(error);
      return response;
    }


  }


  async getUserByIdHandler(request, h) {
    try {
      const { id } = request.params;

      const user = await this._service.getUserById(id);

      return {
        status: "success",
        data: {
          user,
        },
      };
    } catch (error) {
      if (error instanceof ClientError) {
        const response = h.response({
          status: "fail",
          message: error.message,
        });
        response.code(error.statusCode);
        return response;
      }

      // server ERROR!
      const response = h.response({
        status: "error",
        message: "Maaf, terjadi kegagalan pada server kami.",
      });
      response.code(500);
      console.error(error);
      return response;
    }
  }


  async getUsersByUsernameHandler(request, h) {
    try {
      const { username = "" } = request.query;
      const users = await this._service.getUsersByUsername(username);
      return {
        status: "success",
        data: {
          users,
        },
      };
    } catch (error) {
      if (error instanceof ClientError) {
        const response = h.response({
          status: "fail",
          message: error.message,
        });
        response.code(error.statusCode);
        return response;
      }

      // Server ERROR!
      const response = h.response({
        status: "error",
        message: "Maaf, terjadi kegagalan pada server kami.",
      });
      response.code(500);
      console.error(error);
      return response;
    }
  }
}

module.exports = CompaniesHandler;
